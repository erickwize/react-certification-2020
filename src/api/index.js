export { loginApi } from './login.api';
export { fetchSearchVideos } from './youtube.api';
