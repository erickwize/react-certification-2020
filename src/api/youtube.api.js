import { API_KEY, API_URL } from '../utils/constants';

async function fetchSearchVideos(serach = 'wizeline') {
  try {
    let queryParams = '?part=snippet&maxResults=21';
    queryParams = `${queryParams}&q=${serach}`;

    const response = await fetch(`${API_URL}search${queryParams}&key=${API_KEY}`, {
      method: 'get',
      headers: { 'Content-Type': 'application/json' },
    });
    const data = await response.json();
    return data;
  } catch (err) {
    console.error(err);
  }
}

/* async function fetchComments(videoId) {
  try {
    let queryParams = '?part=id%2Csnippet%2Creplies&order=relevance';
    queryParams = `${queryParams}&videoId=${videoId}`;
    const response = await fetch(
      `${API_URL}commentThreads${queryParams}&key=${API_KEY}`,
      {
        method: 'get',
        headers: { 'Content-Type': 'application/json' },
      }
    );
    const data = await response.json();
    return data;
  } catch (e) {
    console.error(e);
    return e;
  }
} */

export { fetchSearchVideos };
