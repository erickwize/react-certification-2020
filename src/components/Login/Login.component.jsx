import React, { useState } from 'react';
import { useHistory } from 'react-router';

import { useGlobal } from '../../providers/Global.provider';
import { Container, LoginCard, CardTitle, CloseButton, Form, FormInput, Label, InfoContainer, Button } from './Login.styles.js';
import { ProfileSVG } from '../../svg/Profile';
import { PasswordSVG } from '../../svg/Password';
import loginApi from '../../providers/login.api';

export const Login = () =>{
  const history = useHistory();
  const [loginData, setLoginData] = useState({username:'', password:''});
  const [loginProcess, setLoginProcess] = useState({loading:false, error:''});
  const { state, dispatch } = useGlobal();
  const { username, password } = loginData;

  function authenticate(event) {
    event.preventDefault();
    setLoginProcess({...loginProcess, loading:true});
    loginApi(username, password).then((response)=>{
      const data = {...response, authenticated: true}
      dispatch({type:'login', value:data})
      history.goBack();
      setLoginProcess({...loginProcess, loading:false});
    }, (error)=>{
      setLoginProcess({loading: false, error:error.message});
    });
  }

  const handleUsrChange = (e) =>{
    setLoginData({...loginData, username:e.target.value});
  };

  const handlePassChange = (e) =>{
    setLoginData({...loginData, password:e.target.value});
  }

  const goBack = () => {
    history.goBack();
  };
  
  return (
    <Container>
      <LoginCard theme={state.theme}>
        <div>
          <CloseButton onClick={goBack}>&times;</CloseButton>
          <CardTitle><h1>Welcome back!</h1></CardTitle>
        </div>
        <Form onSubmit={authenticate} className="login-form">
          <Label theme={state.theme}>
            <ProfileSVG/>
            <FormInput theme={state.theme} value={username} onChange={handleUsrChange} required type="text" data-testid="username" placeholder="Username"/>
          </Label>
          <Label theme={state.theme}>
            <PasswordSVG/>
            <FormInput theme={state.theme} value={password} onChange={handlePassChange} required type="password" data-testid="password" placeholder="Password"/>
          </Label>
          {
            loginProcess.error &&
            <InfoContainer>
              <p>{loginProcess.error}</p>
            </InfoContainer>
          }
          <Button type="submit" whileTap={{scale:0.9}}>{loginProcess.loading ? '...':'Login'}</Button>
        </Form>
      </LoginCard>
    </Container>
  );
};