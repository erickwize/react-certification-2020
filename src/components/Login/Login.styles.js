import { motion } from "framer-motion";
import styled from "styled-components";

export const Container = styled.div`
    background-color: rgba(0,0,0,0.5);
    display: grid;
    height: 100%;
    left: 0;
    place-items: center;
    position: fixed;
    top: 0;
    width: 100%;
    z-index: 1;

    @media(max-width:320px){
        margin-left: 5%;
        width: 90%;
    }
`;

export const LoginCard = styled.div`
    background-color: ${props => props.theme.login.backgroundColor};
    border-radius: 10px;
    box-shadow: ${props => props.theme.login.boxShadow};
    color: ${props => props.theme.login.color};
    display: grid;
    padding: 2rem;
    place-items: center;
`;

export const CloseButton = styled.div`
    color: #aaa;
    display: flex;
    float: right;    
    font-size: 40px;
    font-weight: bold;
    margin: -25px;
    line-height: 25px;
    justify-content: flex-end;
    width: 100%;

    &:hover, &:focus {
        color: black;
        text-decoration: none;
        cursor: pointer
    }
`;

export const CardTitle = styled.div`
    width: 100%;
`;

export const Form = styled.form`
    display: grid;
    place-items: center;
    width: 100%;
`;

export const Label = styled.label`
    width: 100%;
    svg{
        position: absolute;
        margin: 4px 0 0 5px;
        fill: ${props => props.theme.login.svgColor};
    }
`;

export const FormInput = styled.input`
    background-color: ${props =>  props.theme.login.inputBackgroundColor};
    border: none;
    color: ${props =>  props.theme.login.color};
    border-radius: 20px;
    height: 40px;
    margin: 0 0 30px 0;
    outline: none;
    padding: 20px;
    padding-left: 45px;
    width: 100%;
`;

export const InfoContainer = styled.div`
    p{    
        background-color: #f8d7da;
        border-radius:5px;
        color: #721c24;
        font-size: 14px;
        margin-top: -10px;
        padding: 10px;
    }
`;

export const Button = styled(motion.button)`
    background-color: #cd76dc;
    border: none;
    border-radius: 5px;
    color: white;
    cursor: pointer;
    font-size: 17px;
    outline none;
    height: 2.5rem;
    width: 35%;
`;
