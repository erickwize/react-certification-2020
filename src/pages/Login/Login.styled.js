import styled from 'styled-components';

const LoginSecion = styled.section`
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: rgba(0, 0, 0, 0.7);
  z-index: 555;
`;

const Card = styled.div`
  position: fixed;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  background: #f1f1f1;
  z-index: 1000;
  width: 360px;
  height: 340px;
  padding: 10px 20px;
  border-radius: 10px;
`;

const Title = styled.h1`
  text-align: center;
  margin: 5px;
  color: #363537;
`;

const DivInput = styled.div`
  text-align: left;
  margin-top: 10px;
`;

const Label = styled.label`
  display: inline-block;
  margin-bottom: 0.5rem;
  color: #888888;
  font-size: 14px;
`;

const Input = styled.input`
  width: 100%;
  outline: none;
  padding: 8px 16px;
  border: 1px solid #e0e6e8;
  border-radius: 4px;
  font-size: 14px;
  color: #888888;
  transition: box-shadow 0.2s;

  & ::placeholder {
    color: #dedede;
  }
  &:focus {
    box-shadow: 0 0 0 2px rgb(169, 172, 255, 0.5);
  }
`;

const Button = styled.button`
  width: 100%;
  min-width: 64px;
  border: 0;
  border-radius: 4px;
  margin-top: 10px;
  padding: 8px 16px;
  outline: none;
  background-color: #2f8bfd;
  color: #ffffff;
  font-size: 0.875rem;
  font-weight: 500;
  line-height: 1.5;
  letter-spacing: 0.02857rem;
  cursor: pointer;
  transition: all 0.2s;
  &:hover {
    background-color: #0072ff;
  }
`;

const DisplayError = styled.p`
  color: #c52e2c;
  font-size: 14px;
  font-weight: bold;
  margin: 2px 5px;
`;

export { LoginSecion, Card, Title, DivInput, Label, Input, Button, DisplayError };
