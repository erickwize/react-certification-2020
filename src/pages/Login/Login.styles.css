.login {
  width: 300px;
  padding-top: 15px;
  padding-bottom: 15px;
  padding-left: 30px;
  padding-right: 30px;
  display: block;
  margin: auto;
  margin-top: 30px;
  background-color: rgba(0, 0, 0, 0.1);
  border-radius: 20px;
}

.login h1 {
  text-align: center;
  letter-spacing: -1px;
}

.login-form {
  display: flex;
  flex-direction: column;
  align-items: center;
}

.form-group {
  width: 100%;
  display: flex;
  flex-direction: column;
  margin-bottom: 1rem;
}

.form-group strong {
  display: block;
  font-weight: 700;
  text-transform: capitalize;
  margin-bottom: 0.4rem;
}

.form-group input {
  color: white;
  font-size: 1.2rem;
  width: 100%;
  padding: 0.4rem 0.6rem;
  border-radius: 3px;
  border: 1px solid white;
  background-color: rgba(0, 0, 0, 0.1);
}

.login-form button[type='submit'] {
  width: 5rem;
  margin-top: 1rem;
  padding: 0.4rem 0.6rem;
  font-size: 1.2rem;
  border: none;
  border-radius: 3px;
  cursor: pointer;
}

.loginDark {
  background-color: rgba(255, 255, 255, 0.3);
}

.loginDark h1{
  color: rgb(210, 210, 210);
}

.loginDark form{
  color: rgb(210, 210, 210);
}

.loginDark form button{
  color: white;
  background-color: black;
}

.loginDark form button:hover{
  color: white;
  background-color: rgba(255, 255, 255, 0.2);
}

