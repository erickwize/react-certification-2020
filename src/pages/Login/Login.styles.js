import styled from 'styled-components';
import { color, fontSize } from '../../utils/globalStyle';

export const LoginView = styled.div`
  height: 100vh;
`;
export const LoginForm = styled.form`
  margin: 40% auto;
  display: block;
  width: 100%;
  background-color: ${color.lightCoral};
  padding: 15px 18px;
  height: fit-content;
  color: ${color.white};
  font-size: ${fontSize.l};
  text-align: center;
  border-radius: 3px;
`;

export const Inputs = styled.input`
  color: ${color.oceanBlue};
  font-size: ${fontSize.default};
  border: none;
  height: 30px;
  width: 250px;
  &:focus {
    border: none;
    outline: 0;
  }

  @media screen and (max-width: 750px) {
    width: inherit;
  }
`;

export const LabelWrapper = styled.label`
  color: ${color.white};
  margin: 8px 0;
  display: block;
`;

export const Label = styled.p`
  display: block;
  margin: 0 0 3px;
  font-weight: bold;
`;

export const Submit = styled.button`
  height: fit-content;
  width: 100%;
  border-radius: 4px;
  margin: 15px 0;
  color: white;
  padding: 8px 15px;
  font-size: ${fontSize.l};
  display: block;
  background-color: ${color.oceanBlue};
  border: none;
`;
