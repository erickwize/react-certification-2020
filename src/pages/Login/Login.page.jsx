import React, { useState } from 'react';
import { useHistory } from 'react-router-dom';
import { loginApi } from '../../api';

import {
  LoginView,
  LoginForm,
  Submit,
  Inputs,
  Label,
  LabelWrapper,
} from './Login.styles';

function UserLogin({ dispatch }) {
  const history = useHistory();
  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');

  const authenticateUser = async (event) => {
    event.preventDefault();
    if (username && password) {
      dispatch({
        type: 'SET_USER_INFO',
        payload: await loginApi(username, password),
      });
      history.push('/');
    }
  };

  return (
    <LoginView data-testid="user-login">
      <LoginForm onSubmit={authenticateUser}>
        <div className="form-group">
          <LabelWrapper htmlFor="username">
            <Label>Username</Label>
            <Inputs
              required
              type="text"
              id="username"
              placeholder="username"
              value={username}
              onChange={({ target }) => setUsername(target.value)}
            />
          </LabelWrapper>
        </div>
        <div className="form-group">
          <LabelWrapper htmlFor="password">
            <Label>Password</Label>
            <Inputs
              required
              type="password"
              id="password"
              placeholder="password"
              value={password}
              onChange={({ target }) => setPassword(target.value)}
            />
          </LabelWrapper>
        </div>
        <Submit data-testid="user-login-button" type="submit">
          Log in
        </Submit>
      </LoginForm>
    </LoginView>
  );
}

export default UserLogin;
