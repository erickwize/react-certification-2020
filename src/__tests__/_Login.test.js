import React from 'react';
import { cleanup, render, screen, fireEvent, waitFor } from '@testing-library/react';
import { useHistory } from 'react-router-dom';
import Login from '../pages/Login';
import { loginApi } from '../api';

const dispatch = jest.fn();
jest.mock('react-router-dom', () => ({
  useHistory: () => ({
    push: jest.fn(),
  }),
}));

jest.mock('react-router-dom', () => ({
  useHistory: () => ({
    push: jest.fn(),
  }),
}));

afterEach(() => {
  cleanup();
});

it('Should render Login', () => {
  const login = render(<Login />);
  expect(login).toMatchSnapshot();

  const userLogin = screen.getByTestId('user-login');
  expect(userLogin).toBeInTheDocument();
});

it('Login api not called', () => {
  render(<Login dispatch={dispatch} />);

  const userLoginButton = screen.getByTestId('user-login-button');
  expect(userLoginButton).toBeInTheDocument();
  fireEvent.click(userLoginButton);
  expect(dispatch).not.toHaveBeenCalled();
});

it('Fill login data', async () => {
  const spy = jest.spyOn(document, 'getElementById');
  render(<Login dispatch={dispatch} />);

  const username = screen.getByLabelText('Username');
  fireEvent.change(username, { target: { value: 'wizeline' } });
  const password = screen.getByLabelText('Password');
  fireEvent.change(password, { target: { value: 'Rocks!' } });
  const userLoginButton = screen.getByTestId('user-login-button');
  fireEvent.click(userLoginButton);
  expect(username.value).toBe('wizeline');
  expect(password.value).toBe('Rocks!');

  await waitFor(() =>
    expect(dispatch).toHaveBeenCalledWith({
      payload: {
        avatarUrl:
          'https://media.glassdoor.com/sqll/868055/wizeline-squarelogo-1473976610815.png',
        id: '123',
        name: 'Wizeline',
      },
      type: 'SET_USER_INFO',
    })
  );
});

it('Test login api', async () => {
  const result = await loginApi('wizeline', 'Rocks!');
  await waitFor(() =>
    expect(result).toEqual({
      id: '123',
      name: 'Wizeline',
      avatarUrl:
        'https://media.glassdoor.com/sqll/868055/wizeline-squarelogo-1473976610815.png',
    })
  );
});

it('Test login api fail', async () => {
  try {
    await loginApi('wizeline', 'Rock!');
  } catch (e) {
    expect(e.message).toBe('Username or password invalid');
  }
});
