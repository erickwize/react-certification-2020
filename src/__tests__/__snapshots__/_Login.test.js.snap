// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Login 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-bdfBwQ bIQWAJ"
        data-testid="user-login"
      >
        <form
          class="sc-gsTCUz fHuKvs"
        >
          <div
            class="form-group"
          >
            <label
              class="sc-hKgILt hfEduC"
              for="username"
            >
              <p
                class="sc-eCssSg hxWDaq"
              >
                Username
              </p>
              <input
                class="sc-dlfnbm jgFWJL"
                id="username"
                placeholder="username"
                required=""
                type="text"
                value=""
              />
            </label>
          </div>
          <div
            class="form-group"
          >
            <label
              class="sc-hKgILt hfEduC"
              for="password"
            >
              <p
                class="sc-eCssSg hxWDaq"
              >
                Password
              </p>
              <input
                class="sc-dlfnbm jgFWJL"
                id="password"
                placeholder="password"
                required=""
                type="password"
                value=""
              />
            </label>
          </div>
          <button
            class="sc-jSgupP pwIew"
            data-testid="user-login-button"
            type="submit"
          >
            Log in
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-bdfBwQ bIQWAJ"
      data-testid="user-login"
    >
      <form
        class="sc-gsTCUz fHuKvs"
      >
        <div
          class="form-group"
        >
          <label
            class="sc-hKgILt hfEduC"
            for="username"
          >
            <p
              class="sc-eCssSg hxWDaq"
            >
              Username
            </p>
            <input
              class="sc-dlfnbm jgFWJL"
              id="username"
              placeholder="username"
              required=""
              type="text"
              value=""
            />
          </label>
        </div>
        <div
          class="form-group"
        >
          <label
            class="sc-hKgILt hfEduC"
            for="password"
          >
            <p
              class="sc-eCssSg hxWDaq"
            >
              Password
            </p>
            <input
              class="sc-dlfnbm jgFWJL"
              id="password"
              placeholder="password"
              required=""
              type="password"
              value=""
            />
          </label>
        </div>
        <button
          class="sc-jSgupP pwIew"
          data-testid="user-login-button"
          type="submit"
        >
          Log in
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
